app/controllers/contracts_controller.rb

```diff
@@ -36,9 +36,50 @@ def destroy
     redirect_to contracts_path
   end
 
+  def pending
+    @contracts = policy_scope(Contract).where(status: :pending)
+    authorize :contract, :pending?
+  end
+
+  def approve
+    @contract = authorize Contract.find(params["id"])
+
+    if request.patch?
+      supplier = Supplier.find_by(supplier_number: contract_params[:supplier_number])
+      affiliate = Affiliate.find_by(name: contract_params[:affiliate_name])
+
+      if supplier.nil?
+        @contract.errors.add(:supplier_number, "Supplier not found")
+        render :approve and return
+      end
+
+      if affiliate.nil?
+        @contract.errors.add(:affiliate_name, "Affiliate not found")
+        render :approve and return
+      end
+
+      @contract.supplier_id = supplier.id
+      @contract.affiliate_id = affiliate.id
+
```

Comment:
Move the `authorize` to a `authorize @contract` down here to include the state of changing supplier_id and affiliate_id into the policy's consideration.